### PR TITLE
feat: uniswap plugin and example

### PR DESCRIPTION
**Pull request by dannythedawger**
## Description

- This PR aims to introduce the skeleton of a Uniswap plugin
- Lints the repository, hence the diff on a few other plugins
- Adds an example for use of the plugin using *vercel ai*

### CODE REVIEW

File: typescript/packages/plugins/uniswap/src/tools.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        },
+        {
+            name: "send_transaction",
+            description: "This {{tool}} signs and sends a transaction.",
+            parameters: SendTransactionBodySchema,
+            method: async (walletClient: EVMWalletClient, parameters: z.infer<typeof SendTransactionBodySchema>) => {
+                return sendTransaction(parameters, walletClient);
+            },
+        },
+    ];
```
Comment (dannythedawger) [inline typescript/packages/plugins/uniswap/src/tools.ts:48]: had to create this one as I didn't see a wallet plugin that can send a primitive tx

File: typescript/packages/plugins/uniswap/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    baseUrl: string;
+};
+
+const SUPPORTED_CHAINS = [mainnet, polygon, avalanche, base, optimism, zora, arbitrum, celo];
+
+export function uniswap({ apiKey, baseUrl }: UniswapOptions): Plugin<EVMWalletClient> {
+    return {
+        name: "Uniswap",
+        supportsChain: (chain: Chain) => chain.type === "evm" && SUPPORTED_CHAINS.some((c) => c.id === chain.id),
+        supportsSmartWallets: () => false,
```
Comment (0xaguspunk) [inline typescript/packages/plugins/uniswap/src/index.ts:16]: Based on the code it does seem like this plugin would support smart wallets! :)

File: typescript/packages/plugins/uniswap/src/tools.ts (hunk 1 above)
Comment (0xaguspunk) [reply typescript/packages/plugins/uniswap/src/tools.ts:48]: A generic sendTx function is quite complex for an LLM to figure out what to do with it.

The goal with these plugins is to have specific functions (e.g swapTokens, createLiquidityPool, collectFees, etc.)

Then these functions can use sendTransaction internally.

I would remove this one. There is another dev adding those specific calls in a different PR so maybe they could contribute to this base with those functions :)

File: typescript/packages/plugins/uniswap/src/tools.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+export type UniswapToolsOptions = {
+    apiKey: string;
+    baseUrl: string;
+};
+
+export function getTools({ apiKey, baseUrl }: UniswapToolsOptions): DeferredTool<EVMWalletClient>[] {
+    return [
+        {
+            name: "check_approval",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/uniswap/src/tools.ts:14]: There is already the approval action on the erc20 plugin. I would leverage that one and remove this one from here. So I can install both erc20 and uniswap plugins and the agent can use both to achieve their goal

File: typescript/packages/plugins/uniswap/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+## Usage
+
+```typescript
+import { uniswap } from "@goat-sdk/plugin-uniswap";
+
+
+const plugin = uniswap({
+    baseUrl: process.env.UNISWAP_BASE_URL as string,
+    apiKey: process.env.UNISWAP_API_KEY as string,
```
Comment (0xaguspunk) [inline typescript/packages/plugins/uniswap/README.md:18]: Where does one get a Uniswap API key? It'd be great to link to that on the readme

File: typescript/examples/vercel-ai/uniswap/.env.template [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Uniswap (get an API key by running the `pnpm create-api-key` command)
```
Comment (0xaguspunk) [inline typescript/examples/vercel-ai/uniswap/.env.template:1]: Is this script pushed?

**Review by 0xaguspunk — CHANGES_REQUESTED** [review]
Neat! Would love to get this merged! Made a few comments on:
1. Functions we might not need
2. Support for smart wallets
3. Getting an API key to use this

Otherwise, lgtm!!

File: typescript/packages/plugins/uniswap/src/tools.ts (hunk 3 above)
Comment (dannythedawger) [reply typescript/packages/plugins/uniswap/src/tools.ts:14]: this is not a generic approval call, it's part of the Uniswap Trading API suite - it does approvals for all sort of functionality including intents and bridging, with and without Permit2. The plugin is not usable without this.

File: typescript/packages/plugins/uniswap/src/index.ts (hunk 2 above)
Comment (dannythedawger) [reply typescript/packages/plugins/uniswap/src/index.ts:16]: Yep, good callout!

File: typescript/examples/vercel-ai/uniswap/.env.template (hunk 5 above)
Comment (dannythedawger) [reply typescript/examples/vercel-ai/uniswap/.env.template:1]: Nope, my but - I copied it over from a template I used

File: typescript/packages/plugins/uniswap/README.md (hunk 4 above)
Comment (dannythedawger) [reply typescript/packages/plugins/uniswap/README.md:18]: Ah I included the link to the landing page and form in the *vercel ai* example I created - shall I add it here too?

**Comment by dannythedawger** [comment]
@0xaguspunk thansk for great feedback! I addressed most of the comments and left 1 open question to discuss further. We should also introduce lint checks before merging as the bulk of my PR is linting other pkg's code!

**Comment by 0xaguspunk** [comment]
Neat 🔥 Will take a look at it in the morning! Agree with the lint, so annoying, haha will look into that too!

File: typescript/packages/plugins/uniswap/src/tools.ts (hunk 1 above)
Comment (0xaguspunk) [reply typescript/packages/plugins/uniswap/src/tools.ts:48]: That is the case for AI SDK but other AI agent frameworks are not there yet. In any case let's go with this approach! :)

Could you move the sendTransaction method [here](https://github.com/goat-sdk/goat/tree/main/typescript/packages/core/src/plugins) instead? Make it a core plugin instead.

**Review by 0xaguspunk — APPROVED** [review]
This is great 🔥

Last (sorry if I'm asking too much!): maybe rebase from main and run `pnpm change:add` to make sure we release this! :)

**Comment by dannythedawger** [comment]
@0xaguspunk thanks for the approval, addressed your last comment too! For some reason, the tests won't run

**Review by 0xaguspunk — APPROVED** [review]
Two small things:
1. Could we call it `eth-send-transaction.ts` to make it consistent?
2. Does it need to be exported from the `index.ts` file in the root of `core`?

Let's merge it after this!

**Comment by dannythedawger** [comment]
> Two small things:
> 
> 1. Could we call it `eth-send-transaction.ts` to make it consistent?
> 2. Does it need to be exported from the `index.ts` file in the root of `core`?
> 
> Let's merge it after this!

Should be ready to merge - again, the build and test command are not running for some reason

**Comment by 0xaguspunk** [comment]
hmm weird. Will merge and take a look at it

**Pull request merged**